File: skills/asset-validation/scripts/plugin_runtime.py
import json
import shutil
from pathlib import Path
from typing import Optional

from .envprep import prepare_round_environment


def _copy_entry(src: Path, dest: Path) -> None:
    if src.is_dir():
        if dest.exists() or dest.is_symlink():
            if dest.is_dir() and not dest.is_symlink():
                shutil.rmtree(dest)
            else:
                dest.unlink()
        shutil.copytree(src, dest, symlinks=True)
    else:
        dest.parent.mkdir(parents=True, exist_ok=True)
        shutil.copy2(src, dest)


def _install_manifest_path(sandbox) -> Path:
    return Path(sandbox) / ".aut-acceptance" / "plugin-install.json"


def _read_install_manifest(sandbox) -> dict:
    path = _install_manifest_path(sandbox)
    if not path.exists():
        return {}
    try:
        data = json.loads(path.read_text(encoding="utf-8"))
        return data if isinstance(data, dict) else {}
    except (OSError, json.JSONDecodeError):
        return {}


def _write_install_manifest(sandbox, manifest: dict) -> None:
    path = _install_manifest_path(sandbox)
    path.parent.mkdir(parents=True, exist_ok=True)
    path.write_text(json.dumps(manifest, ensure_ascii=False, indent=2),
                    encoding="utf-8")


def _session_isolation_prompt(kind: str, names: list) -> str:
    listed = ", ".join(names) if names else kind
    return (
        "Acceptance sandbox isolation: use only the asset staged through this "
        f"session plugin ({kind}: {listed}). Do not use same-name global skills, "
        "agents, or files under $HOME/.claude/skills as the asset source when a "
        "staged copy exists under ACCEPTANCE_SANDBOX. A token such as "
        "$asset-validation denotes a skill name in the user task, not a shell "
        "command or npm package. Never print settings files, auth tokens, or "
        "secret environment values; report paths only."
    )


def _read_plugin_name(src: Path, fallback: Optional[str]) -> str:
    manifest = src / "plugin.json"
    if manifest.exists():
        try:
            data = json.loads(manifest.read_text(encoding="utf-8"))
            if data.get("name"):
                return data["name"]
        except (OSError, json.JSONDecodeError):
            pass
    return fallback or src.name


def _read_skill_name(src: Path, fallback: Optional[str]) -> str:
    skill = src / "SKILL.md"
    if skill.exists():
        try:
            in_frontmatter = False
            for raw in skill.read_text(encoding="utf-8").splitlines():
                line = raw.strip()
                if line == "---":
                    if not in_frontmatter:
                        in_frontmatter = True
                        continue
                    break
                if in_frontmatter and line.startswith("name:"):
                    name = line.split(":", 1)[1].strip().strip("\"'")
                    if name:
                        return name
        except OSError:
            pass
    return fallback or src.name


def _read_agent_name(src: Path, fallback: Optional[str]) -> str:
    candidates = []
    if src.is_file() and src.suffix == ".md":
        candidates.append(src)
    agents_dir = src / "agents"
    if agents_dir.exists():
        candidates.extend(sorted(child for child in agents_dir.iterdir()
                                 if child.suffix == ".md"))
    for child in candidates:
        try:
            in_frontmatter = False
            for raw in child.read_text(encoding="utf-8").splitlines():
                line = raw.strip()
                if line == "---":
                    if not in_frontmatter:
                        in_frontmatter = True
                        continue
                    break
                if in_frontmatter and line.startswith("name:"):
                    name = line.split(":", 1)[1].strip().strip("\"'")
                    if name:
                        return name
        except OSError:
            pass
    return fallback or src.stem


def install_plugin_source(sandbox, source_path, *, name: Optional[str] = None) -> dict:
    src = Path(source_path)
    if not src.exists():
        return {"installed": False, "reason": f"source not found: {src}"}

    env = prepare_round_environment(sandbox)
    plugin_name = _read_plugin_name(src, name)
    plugin_dir = Path(env["ACCEPTANCE_SANDBOX"]) / ".iso" / "claude-plugins" / plugin_name
    _copy_entry(src, plugin_dir)

    skills = sorted(child.name for child in (src / "skills").iterdir()) \
        if (src / "skills").exists() else []
    agents = sorted(child.name for child in (src / "agents").iterdir()) \
        if (src / "agents").exists() else []
    settings = env["CMDAI_CLAUDE_SETTINGS_PATH"]
    cli_args = [
        "--bare",
        "--append-system-prompt", _session_isolation_prompt("plugin", skills + agents),
        "--settings", settings,
        "--plugin-dir", str(plugin_dir),
    ]
    manifest = {
        "name": plugin_name,
        "source": str(src),
        "plugin_dir": str(plugin_dir),
        "settings": settings,
        "skills": skills,
        "agents": agents,
        "cli_args": cli_args,
    }
    _write_install_manifest(sandbox, manifest)

    return {
        "installed": True,
        "name": plugin_name,
        "plugin_dir": str(plugin_dir),
        "settings": settings,
        "cli_args": cli_args,
        "skills": skills,
        "agents": agents,
    }


def install_agent_source(sandbox, source_path, *, name: Optional[str] = None) -> dict:
    src = Path(source_path)
    if not src.exists():
        return {"installed": False, "reason": f"source not found: {src}"}

    env = prepare_round_environment(sandbox)
    agent_name = _read_agent_name(src, name)
    plugin_name = f"acc-agent-{agent_name}"
    plugin_dir = (
        Path(env["ACCEPTANCE_SANDBOX"])
        / ".iso" / "claude-agent-plugins" / plugin_name
    )
    agents_dest = plugin_dir / "agents"
    if (src / "agents").exists():
        _copy_entry(src / "agents", agents_dest)
    elif src.is_file() and src.suffix == ".md":
        _copy_entry(src, agents_dest / src.name)
    plugin_dir.mkdir(parents=True, exist_ok=True)
    (plugin_dir / "plugin.json").write_text(
        json.dumps({"name": plugin_name}, ensure_ascii=False, indent=2),
        encoding="utf-8",
    )

    agents = sorted(child.name for child in agents_dest.iterdir()) \
        if agents_dest.exists() else []
    settings = env["CMDAI_CLAUDE_SETTINGS_PATH"]
    cli_args = [
        "--bare",
        "--append-system-prompt", _session_isolation_prompt("agent", agents),
        "--settings", settings,
        "--plugin-dir", str(plugin_dir),
    ]
    manifest = {
        "name": plugin_name,
        "source": str(src),
        "plugin_dir": str(plugin_dir),
        "settings": settings,
        "skills": [],
        "agents": agents,
        "cli_args": cli_args,
        "staged_asset_type": "agent",
    }
    _write_install_manifest(sandbox, manifest)

    return {
        "installed": True,
        "name": plugin_name,
        "plugin_dir": str(plugin_dir),
        "settings": settings,
        "cli_args": cli_args,
        "skills": [],
        "agents": agents,
    }


def install_skill_source(sandbox, source_path, *, name: Optional[str] = None) -> dict:
    src = Path(source_path)
    if not src.exists():
        return {"installed": False, "reason": f"source not found: {src}"}

    env = prepare_round_environment(sandbox)
    skill_name = _read_skill_name(src, name)
    plugin_name = f"acc-skill-{skill_name}"
    plugin_dir = (
        Path(env["ACCEPTANCE_SANDBOX"])
        / ".iso" / "claude-skill-plugins" / plugin_name
    )
    skill_dir = plugin_dir / "skills" / skill_name
    _copy_entry(src, skill_dir)
    plugin_dir.mkdir(parents=True, exist_ok=True)
    (plugin_dir / "plugin.json").write_text(
        json.dumps({"name": plugin_name}, ensure_ascii=False, indent=2),
        encoding="utf-8",
    )

    settings = env["CMDAI_CLAUDE_SETTINGS_PATH"]
    cli_args = [
        "--bare",
        "--append-system-prompt", _session_isolation_prompt("skill", [skill_name]),
        "--settings", settings,
        "--plugin-dir", str(plugin_dir),
    ]
    manifest = {
        "name": plugin_name,
        "source": str(src),
        "plugin_dir": str(plugin_dir),
        "settings": settings,
        "skills": [skill_name],
        "agents": [],
        "cli_args": cli_args,
        "staged_asset_type": "skill",
    }
    _write_install_manifest(sandbox, manifest)

    return {
        "installed": True,
        "name": plugin_name,
        "plugin_dir": str(plugin_dir),
        "settings": settings,
        "cli_args": cli_args,
        "skills": [skill_name],
        "agents": [],
    }


def cleanup_plugin_install(sandbox) -> Optional[dict]:
    manifest = _read_install_manifest(sandbox)
    if not manifest:
        return None
    raw_plugin_dir = manifest.get("plugin_dir")
    removed_plugin_dir = False
    plugin_dir = Path(raw_plugin_dir) if raw_plugin_dir else None
    if plugin_dir and plugin_dir.exists():
        if plugin_dir.is_dir() and not plugin_dir.is_symlink():
            shutil.rmtree(plugin_dir)
        else:
            plugin_dir.unlink()
        removed_plugin_dir = True
    try:
        _install_manifest_path(sandbox).unlink()
    except FileNotFoundError:
        pass
    return {
        "name": manifest.get("name"),
        "removed_plugin_dir": removed_plugin_dir,
        "plugin_dir": str(plugin_dir) if plugin_dir else None,
        "skills": manifest.get("skills", []),
        "agents": manifest.get("agents", []),
    }

File: skills/asset-validation/scripts/plugin-runtime.ts
import {
  copyFileSync,
  cpSync,
  existsSync,
  lstatSync,
  mkdirSync,
  readFileSync,
  readdirSync,
  realpathSync,
  rmSync,
  statSync,
  unlinkSync,
  writeFileSync,
} from "node:fs";
import { basename, extname, join, parse } from "node:path";

import { prepareRoundEnvironment } from "./envprep.ts";

export interface PluginInstall {
  installed: boolean;
  name?: string;
  reason?: string;
  plugin_dir?: string;
  settings?: string;
  cli_args?: string[];
  skills?: string[];
  agents?: string[];
  asset_type?: string;
  [key: string]: unknown;
}

function pathExistsOrIsSymlink(path: string): boolean {
  try {
    lstatSync(path);
    return true;
  } catch {
    return false;
  }
}

function removeEntry(path: string): void {
  const stats = lstatSync(path);
  if (stats.isDirectory() && !stats.isSymbolicLink()) {
    rmSync(path, { recursive: true });
  } else {
    unlinkSync(path);
  }
}

function copyEntry(source: string, destination: string): void {
  if (statSync(source).isDirectory()) {
    if (pathExistsOrIsSymlink(destination)) removeEntry(destination);
    const copySource = lstatSync(source).isSymbolicLink() ? realpathSync(source) : source;
    cpSync(copySource, destination, {
      recursive: true,
      force: true,
      dereference: false,
      verbatimSymlinks: true,
    });
  } else {
    mkdirSync(join(destination, ".."), { recursive: true });
    copyFileSync(source, destination);
  }
}

function installManifestPath(sandbox: string): string {
  return join(sandbox, ".aut-acceptance", "plugin-install.json");
}

function readInstallManifest(sandbox: string): Record<string, unknown> {
  const path = installManifestPath(sandbox);
  if (!existsSync(path)) return {};
  try {
    const data = JSON.parse(readFileSync(path, "utf8"));
    return typeof data === "object" && data !== null && !Array.isArray(data)
      ? data as Record<string, unknown>
      : {};
  } catch {
    return {};
  }
}

function writeInstallManifest(sandbox: string, manifest: Record<string, unknown>): void {
  const path = installManifestPath(sandbox);
  mkdirSync(join(path, ".."), { recursive: true });
  writeFileSync(path, JSON.stringify(manifest, null, 2), "utf8");
}

function claudePluginManifestPath(pluginDirectory: string): string {
  return join(pluginDirectory, ".claude-plugin", "plugin.json");
}

function ensureClaudePluginManifest(pluginDirectory: string, pluginName: string): string {
  const destination = claudePluginManifestPath(pluginDirectory);
  if (existsSync(destination)) return destination;

  const manifest: Record<string, unknown> = {
    name: pluginName,
    version: "1.0.0",
    description: `Acceptance-isolated staging plugin for ${pluginName}.`,
    author: { name: "asset-validation" },
  };
  for (const candidate of [
    join(pluginDirectory, "plugin.json"),
    join(pluginDirectory, ".codex-plugin", "plugin.json"),
  ]) {
    if (!existsSync(candidate)) continue;
    try {
      const parsed = JSON.parse(readFileSync(candidate, "utf8"));
      if (typeof parsed === "object" && parsed !== null && !Array.isArray(parsed)) {
        const source = parsed as Record<string, unknown>;
        for (const key of [
          "version", "description", "author", "homepage", "repository", "license", "keywords",
        ]) {
          if (key in source) manifest[key] = source[key];
        }
        break;
      }
    } catch {
      // Fall back to the minimal generated manifest.
    }
  }
  mkdirSync(join(destination, ".."), { recursive: true });
  writeFileSync(destination, JSON.stringify(manifest, null, 2), "utf8");
  return destination;
}

function sessionIsolationPrompt(kind: string, names: string[]): string {
  const listed = names.length ? names.join(", ") : kind;
  return "Acceptance sandbox isolation: use only the asset staged through this "
    + `session plugin (${kind}: ${listed}). Do not use same-name global skills, `
    + "agents, or files under $HOME/.claude/skills as the asset source when a "
    + "staged copy exists under ACCEPTANCE_SANDBOX. A slash token such as "
    + "/asset-validation denotes a Claude skill name in the user task, not a "
    + "shell command or npm package. Never print settings files, auth tokens, or "
    + "secret environment values; report paths only.";
}

function readPluginName(source: string, fallback?: string): string {
  for (const manifest of [
    claudePluginManifestPath(source),
    join(source, ".codex-plugin", "plugin.json"),
    join(source, "plugin.json"),
  ]) {
    if (!existsSync(manifest)) continue;
    try {
      const data = JSON.parse(readFileSync(manifest, "utf8")) as Record<string, unknown>;
      if (data.name) return String(data.name);
    } catch {
      // Try the next supported manifest location.
    }
  }
  return fallback || basename(source);
}

function frontmatterName(path: string): string | null {
  try {
    let inFrontmatter = false;
    for (const raw of readFileSync(path, "utf8").split(/\r?\n/)) {
      const line = raw.trim();
      if (line === "---") {
        if (!inFrontmatter) {
          inFrontmatter = true;
          continue;
        }
        break;
      }
      if (inFrontmatter && line.startsWith("name:")) {
        const name = line.slice(line.indexOf(":") + 1).trim().replace(/^["']|["']$/g, "");
        if (name) return name;
      }
    }
  } catch {
    // Fall through to the caller's fallback.
  }
  return null;
}

function readSkillName(source: string, fallback?: string): string {
  const name = frontmatterName(join(source, "SKILL.md"));
  return name || fallback || basename(source);
}

function readAgentName(source: string, fallback?: string): string {
  const candidates: string[] = [];
  if (existsSync(source) && statSync(source).isFile() && extname(source) === ".md") {
    candidates.push(source);
  }
  const agentsDirectory = join(source, "agents");
  if (existsSync(agentsDirectory)) {
    candidates.push(
      ...readdirSync(agentsDirectory)
        .filter((child) => extname(child) === ".md")
        .sort()
        .map((child) => join(agentsDirectory, child)),
    );
  }
  for (const candidate of candidates) {
    const name = frontmatterName(candidate);
    if (name) return name;
  }
  return fallback || parse(source).name;
}

function childNames(path: string): string[] {
  return existsSync(path) ? readdirSync(path).sort() : [];
}

export function installPluginSource(
  sandbox: string,
  sourcePath: string,
  options: { name?: string } = {},
): PluginInstall {
  if (!existsSync(sourcePath)) {
    return { installed: false, reason: `source not found: ${sourcePath}` };
  }
  const env = prepareRoundEnvironment(sandbox);
  const pluginName = readPluginName(sourcePath, options.name);
  const pluginDirectory = join(env.ACCEPTANCE_SANDBOX, ".iso", "claude-plugins", pluginName);
  copyEntry(sourcePath, pluginDirectory);
  ensureClaudePluginManifest(pluginDirectory, pluginName);
  const skills = childNames(join(sourcePath, "skills"));
  const agents = childNames(join(sourcePath, "agents"));
  const settings = env.CMDAI_CLAUDE_SETTINGS_PATH;
  const cliArgs = [
    "--bare",
    "--append-system-prompt",
    sessionIsolationPrompt("plugin", [...skills, ...agents]),
    "--settings",
    settings,
    "--plugin-dir",
    pluginDirectory,
  ];
  const manifest = {
    name: pluginName,
    source: sourcePath,
    plugin_dir: pluginDirectory,
    settings,
    skills,
    agents,
    cli_args: cliArgs,
  };
  writeInstallManifest(sandbox, manifest);
  return { installed: true, ...manifest };
}

export function installAgentSource(
  sandbox: string,
  sourcePath: string,
  options: { name?: string } = {},
): PluginInstall {
  if (!existsSync(sourcePath)) {
    return { installed: false, reason: `source not found: ${sourcePath}` };
  }
  const env = prepareRoundEnvironment(sandbox);
  const agentName = readAgentName(sourcePath, options.name);
  const pluginName = `acc-agent-${agentName}`;
  const pluginDirectory = join(
    env.ACCEPTANCE_SANDBOX,
    ".iso",
    "claude-agent-plugins",
    pluginName,
  );
  const agentsDestination = join(pluginDirectory, "agents");
  if (existsSync(join(sourcePath, "agents"))) {
    copyEntry(join(sourcePath, "agents"), agentsDestination);
  } else if (statSync(sourcePath).isFile() && extname(sourcePath) === ".md") {
    copyEntry(sourcePath, join(agentsDestination, basename(sourcePath)));
  }
  mkdirSync(pluginDirectory, { recursive: true });
  ensureClaudePluginManifest(pluginDirectory, pluginName);
  const agents = childNames(agentsDestination);
  const settings = env.CMDAI_CLAUDE_SETTINGS_PATH;
  const cliArgs = [
    "--bare",
    "--append-system-prompt",
    sessionIsolationPrompt("agent", agents),
    "--settings",
    settings,
    "--plugin-dir",
    pluginDirectory,
  ];
  const manifest = {
    name: pluginName,
    source: sourcePath,
    plugin_dir: pluginDirectory,
    settings,
    skills: [] as string[],
    agents,
    cli_args: cliArgs,
    staged_asset_type: "agent",
  };
  writeInstallManifest(sandbox, manifest);
  return {
    installed: true,
    name: pluginName,
    plugin_dir: pluginDirectory,
    settings,
    cli_args: cliArgs,
    skills: [],
    agents,
  };
}

export function installSkillSource(
  sandbox: string,
  sourcePath: string,
  options: { name?: string } = {},
): PluginInstall {
  if (!existsSync(sourcePath)) {
    return { installed: false, reason: `source not found: ${sourcePath}` };
  }
  const env = prepareRoundEnvironment(sandbox);
  const skillName = readSkillName(sourcePath, options.name);
  const pluginName = `acc-skill-${skillName}`;
  const pluginDirectory = join(
    env.ACCEPTANCE_SANDBOX,
    ".iso",
    "claude-skill-plugins",
    pluginName,
  );
  const skillDirectory = join(pluginDirectory, "skills", skillName);
  copyEntry(sourcePath, skillDirectory);
  mkdirSync(pluginDirectory, { recursive: true });
  ensureClaudePluginManifest(pluginDirectory, pluginName);
  const settings = env.CMDAI_CLAUDE_SETTINGS_PATH;
  const cliArgs = [
    "--bare",
    "--append-system-prompt",
    sessionIsolationPrompt("skill", [skillName]),
    "--settings",
    settings,
    "--plugin-dir",
    pluginDirectory,
  ];
  const manifest = {
    name: pluginName,
    source: sourcePath,
    plugin_dir: pluginDirectory,
    settings,
    skills: [skillName],
    agents: [] as string[],
    cli_args: cliArgs,
    staged_asset_type: "skill",
  };
  writeInstallManifest(sandbox, manifest);
  return {
    installed: true,
    name: pluginName,
    plugin_dir: pluginDirectory,
    settings,
    cli_args: cliArgs,
    skills: [skillName],
    agents: [],
  };
}

export function installCodexSkillSource(
  sandbox: string,
  sourcePath: string,
  options: { name?: string } = {},
): PluginInstall {
  if (!existsSync(sourcePath)) {
    return { installed: false, reason: `source not found: ${sourcePath}` };
  }
  const env = prepareRoundEnvironment(sandbox);
  const skillName = readSkillName(sourcePath, options.name);
  const skillDirectory = join(
    env.ACCEPTANCE_SANDBOX,
    ".agents",
    "skills",
    skillName,
  );
  copyEntry(sourcePath, skillDirectory);
  const sourceSkill = statSync(sourcePath).isDirectory()
    ? join(sourcePath, "SKILL.md")
    : sourcePath;
  const cliArgs = [
    "-c",
    `skills.config=[{path=${JSON.stringify(sourceSkill)},enabled=false}]`,
  ];
  const manifest = {
    name: skillName,
    source: sourcePath,
    plugin_dir: skillDirectory,
    skill_dir: skillDirectory,
    settings: null,
    skills: [skillName],
    agents: [] as string[],
    cli_args: cliArgs,
    staged_asset_type: "skill",
    staging: "codex-repo-skill",
  };
  writeInstallManifest(sandbox, manifest);
  return { installed: true, ...manifest };
}

export function cleanupPluginInstall(sandbox: string): Record<string, unknown> | null {
  const manifest = readInstallManifest(sandbox);
  if (!Object.keys(manifest).length) return null;
  const rawPluginDirectory = manifest.plugin_dir;
  const pluginDirectory = rawPluginDirectory ? String(rawPluginDirectory) : null;
  let removedPluginDirectory = false;
  if (pluginDirectory && existsSync(pluginDirectory)) {
    removeEntry(pluginDirectory);
    removedPluginDirectory = true;
  }
  try {
    unlinkSync(installManifestPath(sandbox));
  } catch (error) {
    if ((error as NodeJS.ErrnoException).code !== "ENOENT") throw error;
  }
  return {
    name: manifest.name ?? null,
    removed_plugin_dir: removedPluginDirectory,
    plugin_dir: pluginDirectory,
    skills: manifest.skills ?? [],
    agents: manifest.agents ?? [],
  };
}
